os/kernel/task_monitor/task_monitor.c

```diff
@@ -37,27 +85,59 @@ void task_monitor_update_list(int pid, int status)
 int task_monitor(int argc, char *argv[])
 {
 	int pid_idx;
+	int time_idx;
+	int time_cnt;
+	int max_time;
 	struct tcb_s *tcb;
+	taks_monitor_t *next_mon;
+
+	max_time = CONFIG_TASK_MONITOR_MAX_INTERVAL / CONFIG_TASK_MONITOR_INTERVAL;
+	/* Initialize the dequeue list */
+	for(time_idx = 0; time_idx < max_time; ++time_idx) {
+		dq_init(&g_time[time_idx]);
+	}
 
 	/* Initialize the monitored tasks list */
 	for (pid_idx = 0; pid_idx < CONFIG_MAX_TASKS; pid_idx++) {
-		monitored_tasks_list[pid_idx] = NOT_MONITORED;
+		monitored_tasks_list[pid_idx].flink = NULL;
+		monitored_tasks_list[pid_idx].blink = NULL;
+		monitored_tasks_list[pid_idx].pid = 0;
+		monitored_tasks_list[pid_idx].time = 0;
 	}
 
+	g_monitor_cnt = 0;
+	time_cnt = 0;
+
+	sem_init(&stop_sem, 0, 0);
+
 	while (1) {
+		if (g_monitor_cnt == 0) {
+			/* if g_monitor_cnt is 0,
+			 * task_monitor is not working.
+			 */
+			sem_wait(&stop_sem);
```

Comment:
This can be awakened by signal. How about checking the return of sem_wait and re-wait?